Src/Support/Google.Apis.Auth/OAuth2/Flows/PkceGoogleAuthorizationCodeFlow.cs

```diff
@@ -0,0 +1,99 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using Google.Apis.Auth.OAuth2.Requests;
+using Google.Apis.Auth.OAuth2.Responses;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Auth.OAuth2.Flows
+{
+    /// <summary>
+    /// Google authorization flow implementation that supports PKCE as described in https://www.rfc-editor.org/rfc/rfc7636
+    /// and https://developers.google.com/identity/protocols/oauth2/native-app.
+    /// </summary>
+    public class PkceGoogleAuthorizationCodeFlow : GoogleAuthorizationCodeFlow, IPkceAuthorizationCodeFlow
+    {
+        private const int CodeVerifierMaxLengthChars = 128;
+        // We generate the random bytes and then encode them to Base64 whose characters are 6 bits long.
+        private const int CodeVerifierRandomNumberLengthBytes = CodeVerifierMaxLengthChars * 6 / 8;
+        private const string ChallengeMethodSha256 = "S256";
+        private const string ChallengeMethodPlain = "plain";
+
+        /// <summary>
+        /// Creates a new instance from the given initializer.
+        /// </summary>
+        public PkceGoogleAuthorizationCodeFlow(Initializer initializer) : base(initializer)
+        {
+        }
+
+        /// <inheritdoc/>
+        public AuthorizationCodeRequestUrl CreateAuthorizationCodeRequest(string redirectUri, out string codeVerifier)
+        {
+            var request = CreateAuthorizationCodeRequest(redirectUri) as GoogleAuthorizationCodeRequestUrl;
+            codeVerifier = GenerateCodeVerifier();
+            AddCodeChallenge(request, codeVerifier);
+            return request;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TokenResponse> ExchangeCodeForTokenAsync(string userId, string code, string codeVerifier, string redirectUri, CancellationToken taskCancellationToken)
+        {
+            var request = CreateAuthorizationCodeTokenRequest(userId, code, redirectUri);
+            request.CodeVerifier = codeVerifier;
+
+            return await ExchangeCodeForTokenAsync(userId, request, taskCancellationToken).ConfigureAwait(false);
+        }
+
+        private static string GenerateCodeVerifier()
+        {
+            byte[] data = new byte[CodeVerifierRandomNumberLengthBytes];
+            using RandomNumberGenerator randomNumberGenerator = RandomNumberGenerator.Create();
+            randomNumberGenerator.GetBytes(data);
+
+            // The code verifier alphabet does not allow +, / or =, so we Base64URL encode.
```

Comment:
I think it's more that base64-URL-encoding is simply how the code verifier is specified in the RFC.

Reply:
Yes, true, it just doesn't say so anywhere in our docs, just what the alphabet is, so I unconsciously followed along. Changing.